src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+        }
+
+        public async Task<string> GetTempPEDescriptionXmlAsync(string fileName)
+        {
+            await InitializeAsync();
+
+            await _unconfiguredProjectServices.ThreadingService.SwitchToUIThread();
+
+            DesignTimeInputsItem inputs = AppliedValue.Value;
+            if (fileName == null)
+                throw new ArgumentException("Must supply a file to build", nameof(fileName));
+
+            if (!inputs.Inputs.Contains(fileName))
+                throw new ArgumentException("FileName supplied must be one of the DesignTime source files", nameof(fileName));
+
+            string outputPath = inputs.OutputPath;
+
+            HashSet<string> files = GetFilesToCompile(fileName, inputs.SharedInputs);
+            string outputFileName = GetOutputFileName(outputPath, fileName);
+            if (CompilationNeeded(files, outputFileName))
+            {
+                // For parity with legacy we don't care about the compilation result: Legacy only errors here if it runs out of memory queuing the compilation
+                await CompileTempPEAsync(files, outputFileName);
+            }
+
+            // VSTypeResolutionService is the only consumer, and it only uses the codebase element so it's fine to default most of these (VC++ does the same)
+            return $@"<root>
+  <Application private_binpath = ""{outputPath}""/>
+  <Assembly
+    codebase = ""{outputFileName}""
+    name = ""{fileName}""
+    version = ""0.0.0.0""
+    snapshot_id = ""1""
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        internal bool CompilationNeeded(HashSet<string> files, string outputFileName)
+        {
+            if (!_fileSystem.FileExists(outputFileName))
+                return true;
+
+            DateTime outputDateTime = _fileSystem.LastFileWriteTime(outputFileName);
+
+            foreach (string file in files)
+            {
+                DateTime fileDateTime = _fileSystem.LastFileWriteTime(file);
+                if (fileDateTime > outputDateTime)
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static string GetOutputFileName(string outputPath, string fileName)
+        {
+            return Path.Combine(outputPath, fileName.Replace('\\', '.') + ".dll");
+        }
+
+        protected virtual Task CompileTempPEAsync(HashSet<string> filesToCompile, string outputFileName)
+        {
+            return _sequentialTaskQueue.ExecuteTask(async () =>
+            {
+                bool result = await _compiler.CompileAsync(_languageServiceHost.ActiveProjectContext, outputFileName, filesToCompile, default);
+
+                // if the compilation failed or was cancelled we should clean up any old TempPE outputs lest a designer gets the wrong types, plus its what legacy did
+                if (!result)
+                {
+                    try
+                    {
+                        if (_fileSystem.FileExists(outputFileName))
```

Comment:
RemoveFile already does this.

Reply:
I'll change it in this case, since I'm just catching all exceptions anyway, but this feels like coding to an implementation detail, which I don't think is normally a good idea.